code/obj/item/gun/energy.dm

```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
+				user.put_in_hand_or_drop(our_light)
+				our_light = null
+				update_icon()
+				return
+			else if (istype(W, /obj/item/light/tube) && !our_light)
+				boutput(user,SPAN_NOTICE("You place [W] inside of the barrel and redo the wiring."))
+				playsound(src, 'sound/effects/pop.ogg', 50, TRUE)
+				attach_light(W, user)
+				update_icon()
+				return
+			..()
+		else
+			boutput(user,SPAN_NOTICE("Attempting to work on [src] while its on fire might be a bad idea..."))
+			return
+
+	get_desc()
+		. = ..()
+		if (!heat_repair)
+			if (!our_cell)
+				. += SPAN_ALERT("<b> [src] is lacking a power source!</b>")
+			if (!our_light)
+				. += SPAN_ALERT("<b> [src] is lacking a light source!</b>")
+			else if(our_light.light_status != LIGHT_OK)
+				. += SPAN_ALERT("<b> [src]'s light source is nonfunctional!</b>")
+		else
+			. += SPAN_ALERT("<b> [src] is broken and requires repairs!</b>")
+
+	get_help_message(dist, mob/user)
+		switch(src.heat_repair)
+			if(0)
+				if(!our_cell)
+					. += "You can use a large energy cell on [src] to attach it to the gun."
+				else
+					. += "You can use a <b>wrench</b> to remove [src]'s energy cell."
+				if(!our_light)
+					. += "You can use a light tube on [src] to insert it into the gun."
+				else
+					. += "You can use <b>wirecutters</b> to remove [src]'s light tube."
+			if(1)
+				. = "You can use <b>wirecutters</b> to remove the burnt wiring."
+			if(2)
+				. = "You can add 10 wire to replace the wiring."
+
+	attack_self(mob/user)
+		var/I = tgui_input_number(user, "Input a firerate (In deciseconds)", "Timer Adjustment", shoot_delay, 10, 2)
+		if (!I || BOUNDS_DIST(src, user) > 0)
+			return
+		shoot_delay = I
+		boutput(user, SPAN_NOTICE("You adjust [src] to fire every [I / 10] seconds."))
+
+	update_icon()
+		if (our_cell)
+			var/image/overlay_image
+			if (istype(our_cell, /obj/item/cell/artifact))
+				var/obj/item/cell/artifact/C = our_cell
+				var/datum/artifact/powercell/AS = C.artifact
+				var/datum/artifact_origin/AO = AS.artitype
+				overlay_image = SafeGetOverlayImage("gun_cell", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-[AO.name]")
+			else
+				overlay_image = SafeGetOverlayImage("gun_cell", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-[our_cell.icon_state]")
+			src.UpdateOverlays(overlay_image, "gun_cell")
+		else
+			src.UpdateOverlays(null, "gun_cell")
+
+		if (our_light)
+			var/image/overlay_image = SafeGetOverlayImage("gun_light", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-light")
+			src.UpdateOverlays(overlay_image, "gun_light")
+		else
+			src.UpdateOverlays(null, "gun_light")
+
+		if (heat > FIRE_THRESHOLD)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-burn")
+			src.UpdateOverlays(overlay_image, "gun_smoke")
+		else if (heat > 70)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-smoke")
```

Comment:
Instances of `'icons/obj/items/guns/energy64x32.dmi'` can be replaced by src.icon